src/v/cluster/tm_stm.h

```diff
@@ -167,10 +171,14 @@ class tm_stm final : public persisted_stm {
 
     std::vector<kafka::transactional_id> get_expired_txs();
 
+protected:
+    ss::future<> handle_eviction() override;
+
 private:
     ss::future<> apply_snapshot(stm_snapshot_header, iobuf&&) override;
     ss::future<stm_snapshot> take_snapshot() override;
 
+    ss::basic_rwlock<> _state_lock;
```

Comment:
all the locking looks correct.

would it be easier for devs to reason about the locking if the tm_stm required external synchronization (or internal, but providing external looks easier in the case of tm_stm: I see rm_stm provides internal already).

right now there is mixed internal/external synchronization for tm_stm: users get a write_lock automatically for handle_eviction, but then have to remember manage read locks externally across all interactions with tm_stm that read (hidden) internal state.

it looks like tm_stm could be treated by the frontend as a piece of state with external synchronization provided by
by (1) moving the state_lock into the tx_gateway_frontend, and (2) taking the write_lock explicitly around handle_eviction calls.

Reply:
@rystsov to be clear this is an optional thing. i think it would be cleaner, but it doesn't need to block merging. let me know

Reply:
Yeah, it's tricky to get right because we should take write lock inside tm_stm (handle_log_eviction is its virtual methos) but we need to take read lock at the tx_gateway_frontend level (it should orchestrate multiple to tm_stm and rm_stm under it).